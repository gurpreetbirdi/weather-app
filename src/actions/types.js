// action types for redux
export const TYPES = {
  WEATHER_INFO: 'WEATHER_INFO',
  LOADER: 'LOADER',
};
