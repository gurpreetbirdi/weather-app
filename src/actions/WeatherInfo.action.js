import { API_BASE_URL, END_POINTS, APP_ID } from '../config/APIURI';
import ApiCall from '../middleware/ApiCall';
import { TYPES } from './types';
import { dayWiseAvgTemp } from '../utils/utils';

const setWeatherInfo = (type, payload) => ({ type, payload });

const setLoader = (type, payload) => ({ type, payload });

export const setUnit = (type, payload) => ({ type, payload });

export const setSelectedCard = (type, payload = '') => ({ type, payload });

// fetch the weather info for 5 days & every 3 hours
export const getWeatherInfo = ({ location = 'Munich,de', units }) => async dispatch => {
  dispatch(setLoader(TYPES.LOADER, true));
  const config = {
    url:
      API_BASE_URL +
      END_POINTS.GET_WEATHER_INFO.replace('{LOCATION}', location).replace('{APP_ID}', APP_ID).replace('{UNITS}', units),
  };
  try {
    const weatherInfoResponse = await ApiCall.getCall(config);
    if (weatherInfoResponse?.data && weatherInfoResponse?.status === 200) {
      weatherInfoResponse.data.dayWiseList = dayWiseAvgTemp(weatherInfoResponse.data.list);
      await dispatch(setWeatherInfo(TYPES.WEATHER_INFO, weatherInfoResponse.data));
    } else {
      await dispatch(setWeatherInfo(TYPES.WEATHER_INFO, {}));
    }
    dispatch(setLoader(TYPES.LOADER, false));
    return weatherInfoResponse;
  } catch (error) {
    dispatch(setLoader(TYPES.LOADER, false));
    return error;
  }
};
