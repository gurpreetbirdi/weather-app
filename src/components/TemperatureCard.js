import React from 'react';
import PropTypes from 'prop-types';
import { Grid, Card, CardContent, Typography, CardActions } from '@material-ui/core';
import { makeStyles } from '@material-ui/core/styles';
import clsx from 'clsx';
import { useSelector, useDispatch } from 'react-redux';

import { constants } from '../config/constants';
import { setSelectedCard } from '../actions/WeatherInfo.action';

const useStyles = makeStyles(() => ({
  alignment: { alignItems: 'center' },
  margin: { margin: '0.1em', padding: '0.1em' },
  card: { margin: '0.2em', width: '33%' },
  boxShadow: {
    boxShadow: '0 4px 8px 0 rgba(0, 0, 0, 0.2), 0 6px 20px 0 rgba(0, 0, 0, 0.19)',
  },
  selectedBox: { boxShadow: '2px 5px 3px rgb(117, 178, 238)' },
  zeroPadding: { padding: 0 },
  noWrap: { whiteSpace: 'nowrap' },
}));

const TemperatureCard = ({ mouseHover, data, setMouseHover }) => {
  const classes = useStyles();
  const dispatch = useDispatch();

  const { unit, selectedCard } = useSelector(state => state.weather);
  const unitValue = unit === constants.FAHRENHEIT ? 'F' : 'C';

  return (
    <Grid
      item
      className={clsx(classes.card, {
        [classes.boxShadow]: mouseHover === data.date,
        [classes.selectedBox]: selectedCard === data.date,
      })}
      onMouseEnter={() => setMouseHover(data.date)}
      onMouseLeave={() => setMouseHover(null)}
      onClick={() => dispatch(setSelectedCard(data.date))}>
      <Card variant="outlined">
        <CardActions>
          <CardContent className={classes.zeroPadding}>
            <Grid container className={classes.alignment}>
              <Typography variant="h5" component="h2" className={classes.margin}>
                Temp:
              </Typography>
              <Typography variant="h6" component="h2" className={classes.margin} color="textSecondary">
                {data.avgTemp} &deg;{unitValue}
              </Typography>
            </Grid>
            <Grid container className={classes.alignment}>
              <Typography variant="h5" component="h2" className={classes.margin}>
                Date:
              </Typography>
              <Typography
                variant="h6"
                component="h2"
                className={clsx(classes.margin, classes.noWrap)}
                color="textSecondary">
                {data.date}
              </Typography>
            </Grid>
            <Grid container className={classes.alignment}>
              <Typography variant="h5" component="h2" className={classes.margin}>
                ...
              </Typography>
            </Grid>
          </CardContent>
        </CardActions>
      </Card>
    </Grid>
  );
};

TemperatureCard.propTypes = {
  mouseHover: PropTypes.string,
  data: PropTypes.shape({
    avgTemp: PropTypes.string,
    date: PropTypes.string,
  }),
  setMouseHover: PropTypes.func,
};

export default TemperatureCard;
