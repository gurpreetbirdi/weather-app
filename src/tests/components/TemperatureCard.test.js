import React from 'react';
import { shallow, mount } from 'enzyme';
import { Provider } from 'react-redux';

import TemperatureCard from '../../components/TemperatureCard';
import { store } from '../store/mockStore';
import { setSelectedCard } from '../../actions/WeatherInfo.action';
import { TYPES } from '../../actions/types';

const data = {
  avgTemp: '57.43',
  date: '06 Jun 20',
};

test('should render TemperatureCard correctly', () => {
  const wrapper = shallow(
    <Provider store={store}>
      <TemperatureCard data={data} />
    </Provider>,
  );
  expect(wrapper.childAt(0)).toMatchSnapshot();
});

describe('<TemperatureCard />', () => {
  const setMouseHover = jest.fn();

  const wrapper = mount(
    <Provider store={store}>
      <TemperatureCard data={data} setMouseHover={setMouseHover} />
    </Provider>,
  );

  test('should call setMouseHover on mouse enter with date', () => {
    wrapper.find('#temp-grid').first().simulate('mouseenter');
    expect(setMouseHover).toHaveBeenCalledWith(data.date);
  });

  test('should call setMouseHover on mouse leave with null', () => {
    wrapper.find('#temp-grid').first().simulate('mouseleave');
    expect(setMouseHover).toHaveBeenCalledWith('');
  });

  test('should dispatch setSelectedCard on click with date', () => {
    wrapper.find('#temp-grid').first().simulate('click');
    expect(store.dispatch(setSelectedCard(data.date))).toEqual({
      type: TYPES.SET_SELECTED_CARD,
      payload: data.date,
    });
  });
});
