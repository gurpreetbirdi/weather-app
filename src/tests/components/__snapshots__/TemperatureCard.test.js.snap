// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render TemperatureCard correctly 1`] = `
<TemperatureCard
  data={
    Object {
      "avgTemp": "57.43",
      "date": "06 Jun 20",
    }
  }
/>
`;
